Add a protocol to relative urls in FileHandler fixes #4918

octfx opened the pull request:
This PR is made in Reference to #4918 

Tries to add a protocol to a relative url, by checking `wgCanonicalServer` or the request through `RequestContext::getMain`. If both fail `http` is used as a fallback.

octfx committed:
Add a protocol to relative urls in FileHandler fixes #4918

kghbln commented:
@octfx Thanks for providing this patch and please excuse that is has been rotting here since. 

@JeroenDeDauw Could you have a peep at this?

JeroenDeDauw commented:
I'm afraid to merge this without tests

octfx commented:
I'll try to add some when my time permits it!

octfx committed:
Update protocol adding logic

- Add FileIndexerTest

octfx commented:
> I do not think this would work for two reasons:
> 
> * `file_get_contents` also does not understand protocol-less URIs. The URL also needs to be protocolized before it is passed to that.
> * In the function `protocolizeUrl`, you return the unaltered canonical URL if it has a scheme. This means any path or query that was passed to the function would be lost.
> 
> I believe just removing lines 136 to 145 (the `try-catch` block) would fix the second issue. The first issue may be fixed by assigning the protocolized URL to a variable and passing that variable to both `file_get_contents` as well as `get_headers`.

Oh, the `file_get_contents` part is a good catch, that slipped through!

I've updated the try-catch block to add the canonical server protocol to the input url.

Further I've added some tests, but I couldn't get them to run locally as it always errored out with `Use of undefined constant SMW_PHPUNIT_AUTOLOADER_FILE [...] boostrap.php on line 15` :/

octfx committed:
Merge branch 'master' of github.com:SemanticMediaWiki/SemanticMediaWiki

octfx committed:
Merge branch 'master' of github.com:SemanticMediaWiki/SemanticMediaWiki

JeroenDeDauw commented:
@Seb35 does this look good to you?

marijnvanwezel commented:
@JeroenDeDauw Could you take a look at this again when you have the time? We are facing this issue again.

JeroenDeDauw commented:
@marijnvanwezel did you test that this patch fixes the issue?

alistair3149 commented:
Thanks for the patch, @octfx, and apologies it sat open this long.

This is now obsolete. The problem in #4918 (`get_headers(): This function may only be used against URLs` when `$wgServer` is protocol-relative) lived in `FileHandler::fetchContentFromURL()`, which the Elastic file ingest used to fetch attachment content over HTTP.

PR #6138 (commit 2a14189dd, merged August 2025) changed `FileIndexer` to read attachment bytes directly from the MediaWiki file backend via `fetchContentFromFile()` instead of fetching over HTTP. As a result `fetchContentFromURL()`, along with its `get_headers()` and `file_get_contents()` calls, is no longer invoked during ingest, so a protocol-relative URL never reaches PHP's URL functions. The `getFullURL()` value is now only stored as inert `file_path` metadata on the Elasticsearch document and is never dereferenced.

Since the code path this PR patches is no longer reached (and the branch has drifted out of mergeability against the refactored `FileHandler`), I'm closing as obsolete in favour of #6138.

Note: `fetchContentFromURL()` itself still exists with the original implementation. It is currently dead code, but if it is ever wired back into a live path it would want a fix along these lines, which can be handled separately.
